### ISSUE TITLE
Decrease build time: Don't resolve dependencies at configuration time, particularly geckoView

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/fenix#10360]
_tldr; devs check for dependencies and download uncached ones for every build, even if they're not going to use it, slowing down builds. This bug would change it so we download dependencies only when they're needed for the particular build._

Currently – and by default in Android – we resolve dependencies at configuration time which happens on every build. This means for every build we connect to the dependency repository, ask them if there are latest versions, and sometimes download the latest versions. This is inefficient if our build task doesn't use some dependencies: for example, we ask the repository for test dependencies even though we're only building a release build. In my brief investigations using Gradle Build Scans, on most builds the time we spend connecting the repositories seems negligible but we should double-check that assumption.

However, resolving all of our dependencies is non-trivial in two cases:
1) When we need to download a large dependency: for example, geckoView is updated at least once a day so we'll all download it at least once a day. Additionally, even if I only ever build geckoBeta builds, I still need to download geckoNightly AARs. This may be especially slow now that we're not in offices, the builds take up a lot of local disk space, and hitting our servers costs money
2) Jetify runs on all resolved dependencies every build (it seems to be worse when the gradle daemon starts). So even if I'm not using geckoNightly, I need to run jetify on it which takes a non-trivial amount of time (off the top of my head 20s?). We can mitigate this by removing Jetify, which is blocked by GeckoView moving to androidx: https://bugzilla.mozilla.org/show_bug.cgi?id=1553459

To mitigate 1), we could do as the official Improving the Performance of Gradle Builds says and [avoid resolving dependencies at configuration time](https://guides.gradle.org/performance/#dont_resolve_dependencies_at_configuration_time) and instead resolve them as part of an actual build task, particularly for GeckoView.

**Comment by pocmo** [comment]
> Additionally, even if I only ever build geckoBeta builds, I still need to download geckoNightly AARs.

Are you sure about this? Afaik that shouldn't be the case and we had to jump through quite some hoops in automation to make it download all dependencies for caching.

> Jetify runs on all resolved dependencies every build 

Would like to get rid of this too! However since it is only a one-time task (if the dependencies do not change) it is maybe not that super wasteful.

**Comment by mcomella** [comment]
> > Additionally, even if I only ever build geckoBeta builds, I still need to download geckoNightly AARs.
> 
> Are you sure about this? Afaik that shouldn't be the case and we had to jump through quite some hoops in automation to make it download all dependencies for caching.

Yes, I've seen this on many occasions though I'd be curious if others have seen this too. I think most folks don't notice it because they're building with Android Studio but it's more apparent when you build via the CL.

For example, I just ran:
![Screen Shot 2020-05-04 at 14 48 54](https://user-images.githubusercontent.com/759372/81017054-6ef20180-8e16-11ea-8ffe-12e23a0ee434.png)

(which triggered a new gradle daemon). And here is gradle resolving dependencies for geckoNightly and geckoBeta concurrently:

![Screen Shot 2020-05-04 at 14 48 36](https://user-images.githubusercontent.com/759372/81017087-816c3b00-8e16-11ea-8113-94be6f0fb75a.png)

I already downloaded the dependencies this morning so it's just doing the jetify transform but I've seen it download geckoNightly for beta builds before.

> > Jetify runs on all resolved dependencies every build
> 
> Would like to get rid of this too! However since it is only a one-time task (if the dependencies do not change) it is maybe not that super wasteful.

I've noticed it seems to run more often than it should (i.e. once a day), perhaps related to whether or not I need a new gradle daemon or not. For example, see the build I screenshotted above.

I probably have a different build use case than most devs – command line builds across different build variants – so I might hit this more often than them if it's less likely to happen in other use cases.

**Comment by mcomella** [comment]
And here's a build scan where both `geckoBeta` and `geckoNightly` are resolved in the same build, as are `fennecNightly`, `fenixNightly`, etc.: https://scans.gradle.com/s/3ur7fxa5hcu4c/performance/resolution
